src/stirling/protocol_inference/data_collection.py

```diff
@@ -16,32 +16,44 @@
 
 import os
 import subprocess
-import json
 import argparse
 
 
-def create_pod2pid():
+def create_pod2pid(runtime):
     """
-    Build a mapping of running pod names to host pids.
+    Builds a mapping of running pod names to host pids.
     """
-    # This cmd gets the host pid of each running container and its corresponding labels.
-    list_ps_cmd = '''kubectl node-shell ''' + node + ''' -- bash -c "docker ps -q | \
-    xargs docker inspect --format '{{.State.Pid}}{{json .Config.Labels}}'"'''
+
+    # This command prints one line of output per container where each line is the pid of the container, a
+    # single space and the name of the k8s pod determined by the io.kubernetes.pod.name container label.
+    # See the output below for example output:
+
+    # $ sudo crictl ps -q | sudo xargs crictl inspect --template \
+    #     '{{.info.pid}} {{index .info.config.labels "io.kubernetes.pod.name"}}' -o go-template
+    # 3635260 vizier-pem-sx7pr
+    # 3634678 kelvin-cdf78c57c-qlzlb
+
+    list_ps_cmd = '''kubectl node-shell ''' + node + ''' -- sudo bash -c "crictl ps -q | \
+    xargs crictl inspect -o go-template \
+    --template '{{.info.pid}} {{index .info.config.labels \\"io.kubernetes.pod.name\\"}}'"'''
+    if runtime == 'docker':
+        list_ps_cmd = '''kubectl node-shell ''' + node + ''' -- sudo bash -c "docker ps -q | \
+    xargs docker inspect --format '{{.State.Pid}} {{index .Config.Labels \\"io.kubernetes.pod.name\\"}}'"'''
 
     pod2pid = {}
     ps_out = subprocess.run(list_ps_cmd, shell=True, capture_output=True)
-    rows = str(ps_out.stdout).split('\\n')[1:-3]
+
+    rows = ps_out.stdout.decode('utf-8').splitlines()
     for row in rows:
-        # Example: '3217413{..., "io.kubernetes.pod.name": "vizier-query-broker", ...}'
-        pid_cutoff_idx = row.index("{")
-        pid = row[:pid_cutoff_idx]
-        try:
-            kube_info = json.loads(format(row[pid_cutoff_idx:]).replace("\\\\", "\\"))
-        except RuntimeError:
-            print(f"Failed to parse config for pid: {pid}. Skipping to the next pid.")
+        split_row = row.split()
+
+        if len(split_row) != 2:
+            print(f"Failed to parse config for row: {row}. Skipping to new pid")
             continue
-        pod_name = kube_info['io.kubernetes.pod.name']
+
+        pid, pod_name = split_row
         pod2pid[pod_name] = pid
+
```

Comment:
`pod2pid[pod_name] = int(pid)`?

Reply:
The rest of the program works fine using the pid as a string. I'd prefer to not introduce casting unless it's necessary.